packages/core/src/invoke.ts

```diff
@@ -29,13 +36,19 @@ export function invokeMachine<
   options: { sync?: boolean } = {}
 ): BehaviorCreator<TContext, TEvent> {
   return (ctx, event, { parent, data, _event }) => {
-    let resolvedMachine = isFunction(machine) ? machine(ctx, event) : machine;
-    if (data) {
-      resolvedMachine = resolvedMachine.withContext(
-        mapContext(data, ctx, _event)
-      ) as TMachine;
-    }
-    return createMachineBehavior(resolvedMachine, parent, options);
+    const resolvedContext = data ? mapContext(data, ctx, _event) : undefined;
+    const machineOrDeferredMachine = isFunction(machine)
+      ? () => {
+          const resolvedMachine = machine(ctx, event);
```

Comment:
promise creators are called with a third argument as well, i think we should strive for unifying those signatures